src/nspawn/nspawn.c

```diff
@@ -2493,6 +2510,65 @@ static int outer_child(
         if (r < 0)
                 return r;
 
+        if (arg_pivot_sysroot) {
+                _cleanup_free_ char *prefixed_pivot_sysroot = NULL;
+                _cleanup_free_ char *pivot_tmp = NULL;
+                _cleanup_free_ char *sysroot = NULL;
+
+                /* Pivot arg_pivot_sysroot to / and the existing / to /sysroot.
+                 * This requires a temporary directory, pivot_tmp, which is
+                 * not a child of either.
+                 *
+                 * This is typically used for OSTree-style containers, where
+                 * the root partition contains several sysroots which could be
+                 * run. Normally, one would be chosen by the bootloader and
+                 * pivoted to / by initramfs.
+                 *
+                 * For example, for an OSTree deployment, arg_pivot_sysroot
+                 * would be: /ostree/deploy/$os/deploy/$checksum. Note that this
+                 * code doesn’t do the /var mount which OSTree expects: use
+                 * --bind +/sysroot/ostree/deploy/$os/var:/var for that.
+                 *
+                 * So in the OSTree case, we’ll end up with something like:
+                 *  - directory = /tmp/systemd-nspawn-123456/root
+                 *  - arg_pivot_sysroot = /ostree/deploy/os/deploy/123abc
+                 *  - prefixed_pivot_sysroot =
+                 *       /tmp/systemd-nspawn-123456/root/ostree/deploy/os/deploy/123abc
+                 *  - pivot_tmp = /tmp/systemd-nspawn-123456/tmp
+                 *  - sysroot = /tmp/systemd-nspawn-123456/tmp/sysroot
+                 */
+                prefixed_pivot_sysroot = prefix_root(directory, arg_pivot_sysroot);
+                pivot_tmp = prefix_root(directory, "../tmp");
+                sysroot = prefix_root(pivot_tmp, "/sysroot");
```

Comment:
I’ve changed one of the `prefix_root()` calls to `prefix_roota()`, but two heap-based calls remain in `setup_pivot_root()` because they are appending user-provided paths (`pivot_root_[new|old]`) which could have any length.